BUG: Fix fill violating read-only flag.

panosz opened the pull request:
- See #22922
- `PyArray_FillWithScalar` checks if destination is writeable before attempting to fill it
- A relevant test is added as a method of `TestRegression`

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

panosz committed:
Bug: Fix fill violating read-only flag.

- See #22922
- `PyArray_FillWithScalar` checks if destination is writeable before
  attempting to fill it
- A relevant test is added as a method of `TestRegression`

panosz commented:
@seberg I guess this does the job, although I am not sure if checking whether the destination is writable should be done inside `PyArray_FillWithScalar`.

Also, wonder if `test_fill_readonly` should cover more test cases.

seberg commented on numpy/core/src/multiarray/convert.c:
```diff
@@ -365,6 +365,10 @@ PyArray_FillWithScalar(PyArrayObject *arr, PyObject *obj)
+
+    if (PyArray_FailUnlessWriteable(arr, "assignment destination") < 0) {
+        return -1;
+    }
```
The comment refers to the allocation below, so either put this check above the comment, or move the check after the first block of variable declarations.

seberg commented on numpy/core/tests/test_regression.py:
```diff
@@ -323,6 +323,13 @@ def bfb():
+        # Ticket #22922
+        with pytest.raises(ValueError):
+            a = np.zeros(11)
+            a.setflags(write=False)
```
Please move this out of the context, best to make the test as specific as possible?  Also we usually put `gh-22922` (and maybe "issue" rather than ticket, but it doesn't matter).  "Ticket #..." sounds like it predates github :). 

I would slightly prefer to move the test into `test_multiarray.py` which has a few `fill` related tests.

seberg reviewed (commented):
Thanks, that is the correct fix, yes.  Just some small comments for fixups to make the change a bit prettier.

panosz committed:
MAINT: Fix comment in PyArray_FillWithScalar. Move test to new location.

-- The comment in `PyArray_FillWithScalar` is about the code after the
readonly check, so it was moved to the appropriate location.
-- The `test_fill_readonly` was moved to test_multiarray.TestAttributes.

panosz commented:
@seberg I modified the code as requested (I believe). 
As for the test, I was not sure about which class I should move it to, because it seems that its purpose is more general than what the names of the test classes imply. Eventually I decided to move it to `TestAttributes`, not because it seemed a good choice, but because all alternatives seemed worse.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -403,6 +403,13 @@ def test_fill_struct_array(self):
+        # gh-22922
+        a = np.zeros(11)
+        a.setflags(write=False)
+        with pytest.raises(ValueError):
```
```suggestion
        with pytest.raises(ValueError, match=".*read-only"):
```
Can't hurt to be even more specific, didn't think of it.

seberg committed:
Apply suggestions from code review

seberg commented on numpy/core/src/multiarray/convert.c:
```diff
@@ -359,12 +359,18 @@ PyArray_ToString(PyArrayObject *self, NPY_ORDER order)
      *
      * (The longlong here should help with alignment.)
      */
+
```
```suggestion
```

seberg committed:
Apply suggestions from code review

seberg commented:
Thanks @panosz those were the changes I was looking for exactly.
